### PR TITLE
[#4291] feat(catalog-lakehouse-paimon): Support Paimon table with primary keys in Gravitino

### PR DESCRIPTION
**Pull request by caican00**
### What changes were proposed in this pull request?

Support Paimon table with primary keys in Gravitino

### Why are the changes needed?

Fix:  https://github.com/apache/gravitino/issues/4291

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
New UTs and ITs.

### CODE REVIEW

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java [hunk 1]
```diff
@@ -51,7 +64,25 @@ protected TableOperations newOps() {
...
+    Schema.Builder builder = Schema.newBuilder().comment(comment);
+    if (properties == null) {
+      properties = Maps.newHashMap();
+    }
+    if (partitioning == null) {
+      partitioning = EMPTY_PARTITIONING;
+    }
+
+    Map<String, String> normalizedProperties = new HashMap<>(properties);
+    normalizedProperties.remove(PRIMARY_KEY_IDENTIFIER);
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java:76]: Gravitino has `PRIMARY_KEY` index, could paimon catalog use it?

Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java:77]: imo, the Gravitino `Index` should be designed for `RDBMS`, am i right?
although paimon's primary key is also unique, there are some differences between it and `RDBMS's` primary key, for example: 
1.  Index in `RDBMS` is globally ordered, but paimon's primary key is ordered  in bucket level.
2.  insert a duplicate record will report an error in RDBMS, but now in paimon, if there is a duplicate record, the internal will be implicit deduplication for insert statement.
3. paimon has not supported the clear concept of column Index so far.  If paimon supports it, we can support it later.

Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java:77]: Hi @FANNG1 do you have further comments about this?

File: catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java [hunk 2]
```diff
@@ -201,6 +207,161 @@ void testCreatePaimonTable() {
...
+                      references.length == 1
+                          && references[0] instanceof NamedReference.FieldReference);
+                  NamedReference.FieldReference fieldReference =
+                      (NamedReference.FieldReference) references[0];
+                  return fieldReference.fieldName()[0];
+                })
+            .toArray(String[]::new);
+    Assertions.assertArrayEquals(partitionKeys, loadedPartitionKeys);
+
+    Assertions.assertTrue(paimonCatalogOperations.tableExists(tableIdentifier));
```
Comment (caican00) [inline catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java:360]: will remove this.

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java (hunk 1 above)
Comment (FANNG1) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java:77]: @jerryshao you thought?

Comment (jerryshao) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java:77]: After discuss with @FANNG1 privately, I think it is suitable to use `PRIMARY_INDEX` than a property to define a key here. Though the behavior of Paimon is slightly different from RDBMS, but it is still some sorts of primary key, right? In Gravitino, we only semantically unify them, not in behavior, each implementation can have it's own explanation, like what we defined for types, transforms, distributions, we only define them semantically, but we don't restrict the behavior, each implementation may be different. So this also applies to here.

Still, we can extend the current PRIMARY_INDEX definition to better support Paimon, WDYT? @caican00

Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java:77]: > but it is still some sorts of primary key, right? 
 yeah.

got it, and it sounds reasonable. I'll use `Index` instead

File: catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java (hunk 2 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java]: fixed.

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java (hunk 1 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java]: fixed.

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java [hunk 3]
```diff
@@ -58,20 +70,26 @@ protected TableOperations newOps() {
...
-                  })
-              .collect(Collectors.toList()));
+    Schema.Builder builder = Schema.newBuilder().comment(comment);
+    if (properties == null) {
+      properties = Maps.newHashMap();
+    }
+    if (partitioning == null) {
+      partitioning = EMPTY_PARTITIONING;
     }
+    if (indexes == null) {
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java:80]: we had checked the nullability of indexs already, this seems redundant?

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java [hunk 4]
```diff
@@ -102,6 +121,77 @@ public static Transform[] toGravitinoPartitioning(List<String> partitionKeys) {
...
+            })
+        .collect(Collectors.toList());
+  }
+
+  private List<String> getPrimaryKeysFromIndexes(Index[] indexes) {
+    if (indexes == null || indexes.length == 0) {
+      return Collections.emptyList();
+    }
+
+    Preconditions.checkArgument(
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java:145]: could you move all the check logic in `checkPaimonPrimaryKeys` and invoke it before create `GravitinoPaimonTable`?

File: catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java [hunk 5]
```diff
@@ -389,7 +458,7 @@ void testGravitinoToPaimonTable() {
     Assertions.assertEquals(new IntType().nullable(), paimonTableSchema.fields().get(0).type());
     Assertions.assertEquals(new DateType().nullable(), paimonTableSchema.fields().get(1).type());
     Assertions.assertEquals(
-        new VarCharType(Integer.MAX_VALUE).nullable(), paimonTableSchema.fields().get(2).type());
+        new VarCharType(Integer.MAX_VALUE).notNull(), paimonTableSchema.fields().get(2).type());
```
Comment (FANNG1) [inline catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java:461]: why changing this line?

**Comment by FANNG1** [comment]
LGTM, except minor comments

File: catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java (hunk 5 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java:461]: https://github.com/apache/gravitino/pull/4292/files#diff-ae0fe81137e01313281a32065109a682c249928f6e91aa331f73abcc070f5668R611-R613
I changed col3 to notnull to make it easier to create a primary key.
it is affected by this change

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java (hunk 3 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java:80]: `toPaimonTableSchema` is called in unit test, afraid that passing `null` will cause `NPE`.

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java (hunk 4 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java:145]: this check seems to help better understand `Index primaryKeyIndex = indexes[0];`
WDYT？

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java (hunk 3 above)
Comment (FANNG1) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java:80]: I prefer to change the test code to remove the check

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java (hunk 4 above)
Comment (FANNG1) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java:145]: ok

File: catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java (hunk 5 above)
Comment (FANNG1) [reply catalogs/catalog-lakehouse-paimon/src/test/java/org/apache/gravitino/catalog/lakehouse/paimon/TestGravitinoPaimonTable.java:461]: got it

File: catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java (hunk 3 above)
Comment (caican00) [reply catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/GravitinoPaimonTable.java]: removed this check.

**Comment by caican00** [comment]
image

Tested IT locally.

**Review by FANNG1 — APPROVED** [review]

**Comment by FANNG1** [comment]
merged to main, thanks for your contribution!

**Pull request merged**